### PR TITLE
[SourceKit] Fail requests when an error occurs

### PR DESCRIPTION
**Pull request by DavidGoldman**
Roll forward of #24124 (reverted in #24993) with fix.

Original description:

Previously, requests would fail silently by returning an empty struct
in the response.

With this change, responses will properly report fail with the internal
error.

This allows IDEs to:

- Properly identify errors when they occur
- Choose to display the error to the user or perhaps handle it

### CODE REVIEW

**Comment by jckarter** [comment]
@swift-ci Please test

File: tools/SourceKit/tools/sourcekitd/lib/API/sourcekitdAPI-XPC.cpp [hunk 1]
```diff
@@ -379,8 +379,9 @@ sourcekitd::createErrorRequestInvalid(const char *Description) {
   return CustomXPCData::createErrorRequestInvalid(Description).getXObj();
 }
 sourcekitd_response_t
-sourcekitd::createErrorRequestFailed(const char *Description) {
-  return CustomXPCData::createErrorRequestFailed(Description).getXObj();
+sourcekitd::createErrorRequestFailed(StringRef DescRef) {
+  std::string Desc = DescRef.str();
+  return CustomXPCData::createErrorRequestFailed(Desc.c_str()).getXObj();
```
Comment (akyrtzi) [inline tools/SourceKit/tools/sourcekitd/lib/API/sourcekitdAPI-XPC.cpp]: Could you change `CustomXPCData::createErrorRequestFailed` to accept a `StringRef` instead ?
Also for consistency please change the other functions as well:
* `sourcekitd::createErrorRequestInvalid`
* `sourcekitd::createErrorRequestInterrupted`
* `CustomXPCData::createErrorRequestInvalid`
* `CustomXPCData::createErrorRequestInterrupted`
* `CustomXPCData::createErrorRequestCancelled`

The `CustomXPCData` functions are just calling`createKindAndString`, change that to accept a `StringRef` like this:
```
  static CustomXPCData createKindAndString(Kind K, StringRef Str) {
    llvm::SmallVector<char, 128> Buf;
    Buf.push_back((char)K);
    Buf.append(Str.begin(), Str.end());
    Buf.push_back('\0');
```

**Review by akyrtzi — CHANGES_REQUESTED** [review]

File: tools/SourceKit/tools/sourcekitd/lib/API/sourcekitdAPI-XPC.cpp (hunk 1 above)
Comment (DavidGoldman) [reply tools/SourceKit/tools/sourcekitd/lib/API/sourcekitdAPI-XPC.cpp]: Done

**Comment by benlangmuir** [comment]
@swift-ci please test

**Comment by benlangmuir** [comment]
@swift-ci please asan test

**Review by akyrtzi — APPROVED** [review]

**Comment by benlangmuir** [comment]
@swift-ci please asan test

**Pull request merged**